internal/session.go

```diff
@@ -0,0 +1,499 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package internal
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/pborman/uuid"
+	"go.uber.org/zap"
+)
+
+func init() {
+	RegisterActivityWithOptions(sessionCreationActivity, RegisterActivityOptions{
+		Name: sessionCreationActivityName,
+	})
+	RegisterActivityWithOptions(sessionCompletionActivity, RegisterActivityOptions{
+		Name: sessionCompletionActivityName,
+	})
+}
+
+type (
+	// SessionInfo contains information of a created session. For now, exported
+	// fields are SessionID and HostName.
+	// SessionID is a uuid generated when CreateSession() or RecreateSession()
+	// is called and can be used to uniquely identify a session.
+	// HostName specifies which host is executing the session
+	SessionInfo struct {
+		SessionID    string
+		HostName     string
+		resourceID   string // hide from user for now
+		tasklist     string // resource specific tasklist
+		sessionState sessionState
+		cancelFunc   CancelFunc
+	}
+
+	// SessionOptions specifies metadata for a session.
+	// ExecutionTimeout: required, no default
+	//     Specifies the maximum amount of time the session can run
+	// CreationTimeout: required, no default
+	//     Specfifies how long session creation can take before returning an error
+	SessionOptions struct {
+		ExecutionTimeout time.Duration
+		CreationTimeout  time.Duration
+	}
+
+	// RecreateSessionParams contains information needed to recreate a session on the same worker.
+	// Use SessionInfo.GetRecreateParameter() and pass the returned value to RecreateSession().
+	// The field of this struct is exported so that it can be pass to new run of a workflow and user
+	// should NOT handcraft this object. Always use GetRecreateParameter() and pass the returned value
+	// to the new run.
+	RecreateSessionParams struct {
+		Tasklist string
```

Comment:
I have to export the field so that the struct can be pass to the new run. User should always use GetRecreateParameter() as written in the comment.